File: snail-race-bet/typescript/src/domain/RaceResultProvider.ts
export class Snail {
    constructor(
        public readonly number: number,
        public readonly name: string,
    ) {
    }
}

export class Podium {
    constructor(
        public readonly first: Snail,
        public readonly second: Snail,
        public readonly third: Snail
    ) {
    }
}

export class SnailRace {
    constructor(
        public readonly raceId:number,
        public readonly timestamp:number,
        public readonly podium:Podium
    ) {
    }

}

function sortFromMoreToLessRecent(a: SnailRace, b:SnailRace) {
    return b.timestamp - a.timestamp
}

export class SnailRaces {
    public readonly races: SnailRace[];

    constructor(races:Array<SnailRace>) {
        this.races = new Array(...races).sort(sortFromMoreToLessRecent)
    }

    static withPodium(snailRaces: SnailRaces, datetime: number, podium: Podium) {
        return new SnailRaces([...snailRaces.races, new SnailRace(999, datetime, podium)]);
    }

    static withAdditionalResult(snailRaces: SnailRaces, raceId: number, datetime: number, podium: Podium) {
        return new SnailRaces([...snailRaces.races, new SnailRace(raceId, datetime, podium)]);
    }
}

export interface RaceResultProvider {
    races(): Promise<SnailRaces>
}

File: snail-race-bet/java/src/main/java/snail/race/kata/domain/RaceResultProvider.java
package snail.race.kata.domain;

import java.util.List;

public interface RaceResultProvider {

    record Snail(int number, String name) {
    }

    record Podium(Snail first,
                  Snail second,
                  Snail third) {
    }

    record SnailRace(int raceId,
                     long timestamp,
                     Podium podium) {
    }

    record SnailRaces(List<SnailRace> races) {
    }

    SnailRaces races();
}
